service/matching/matching_engine.go

```diff
@@ -2928,3 +2937,125 @@ func largerBacklogAge(rootBacklogAge *durationpb.Duration, currentPartitionAge *
 	}
 	return currentPartitionAge
 }
+
+func buildRateLimitConfig(update *workflowservice.UpdateTaskQueueConfigRequest_RateLimitUpdate, updateTime *timestamppb.Timestamp) *taskqueuepb.RateLimitConfig {
+	var rateLimit *taskqueuepb.RateLimit
+	if r := update.GetRateLimit(); r != nil {
+		rateLimit = &taskqueuepb.RateLimit{RequestsPerSecond: r.RequestsPerSecond}
+	}
+	return &taskqueuepb.RateLimitConfig{
+		RateLimit: rateLimit,
+		Metadata: &taskqueuepb.ConfigMetadata{
+			Reason:     update.GetReason(),
+			UpdateTime: updateTime,
+		},
+	}
+}
+
+func prepareTaskQueueUserData(
+	tqud *persistencespb.TaskQueueUserData,
+	taskQueueType enumspb.TaskQueueType,
+) *persistencespb.TaskQueueUserData {
+	data := common.CloneProto(tqud)
+	if data == nil {
+		data = &persistencespb.TaskQueueUserData{}
+	}
+	if data.PerType == nil {
+		data.PerType = make(map[int32]*persistencespb.TaskQueueTypeUserData)
+	}
+	tqType := int32(taskQueueType)
+	if data.PerType[tqType] == nil {
+		data.PerType[tqType] = &persistencespb.TaskQueueTypeUserData{}
+	}
+	if data.PerType[tqType].Config == nil {
+		data.PerType[tqType].Config = &taskqueuepb.TaskQueueConfig{}
+	}
+	return data
+}
+
+func constructUpdateTaskQueueConfigResponse(
+	taskQueueUserData *persistencespb.TaskQueueUserData,
+	taskQueueType enumspb.TaskQueueType,
+) *matchingservice.UpdateTaskQueueConfigResponse {
+	if taskQueueUserData.GetPerType() == nil {
+		return &matchingservice.UpdateTaskQueueConfigResponse{}
+	}
+	tqType := int32(taskQueueType)
+	typeUserData, ok := taskQueueUserData.GetPerType()[tqType]
+	if !ok || typeUserData == nil {
+		return &matchingservice.UpdateTaskQueueConfigResponse{}
+	}
+	updateTaskQueueConfig := typeUserData.GetConfig()
+	if updateTaskQueueConfig == nil {
+		return &matchingservice.UpdateTaskQueueConfigResponse{}
+	}
+	queueRateLimit := updateTaskQueueConfig.GetQueueRateLimit()
+	fairnessKeysRateLimitDefault := updateTaskQueueConfig.GetFairnessKeysRateLimitDefault()
+	return &matchingservice.UpdateTaskQueueConfigResponse{
+		UpdatedTaskqueueConfig: &taskqueuepb.TaskQueueConfig{
+			QueueRateLimit:               queueRateLimit,
+			FairnessKeysRateLimitDefault: fairnessKeysRateLimitDefault,
+		},
+	}
```

Comment:
it still looks like we're doing too much work here, can't we simplify the whole thing to this?

```suggestion
	perTypeUserData := taskQueueUserData.GetPerType()[int32(taskQueueType)]
	return &matchingservice.UpdateTaskQueueConfigResponse{
		UpdatedTaskqueueConfig: perTypeUserData.GetConfig(),
	}
```

Reply:
but also, you have two calls to this function, and one of them just pulls out the config itself and throws away the UpdateTaskQueueConfigResponse wrapper. so I think the function should just return the config and the one caller that needs to wrap it in another message can do that itself.

Reply:
Changed the function to return the config only as it is easier to understand as a function and wrapped the response around the UpdateTaskQueueConfigResponse wherever needed.